examples/developer-tools/src/blocks/type.ts

```diff
@@ -0,0 +1,201 @@
+    // Add new inputs.
+    let i = 0;
+    for (i = 0; i < this.typeCount; i++) {
+      if (!this.getInput('TYPE' + i)) {
+        const input = this.appendValueInput('TYPE' + i);
+        // Disallow chaining arrays of types together.
+        input.setCheck('Type');
+        if (i === 0) {
+          input.appendField('any of');
+        }
+      }
+    }
+    // Remove deleted inputs.
+    while (this.getInput('TYPE' + i)) {
+      this.removeInput('TYPE' + i);
+      i++;
+    }
+  },
+};
+
+export const typeGroupContainer = {
+  // Container.
+  init: function () {
+    this.jsonInit({
+      message0: 'add types %1 %2',
+      args0: [{type: 'input_dummy'}, {type: 'input_statement', name: 'STACK'}],
+      colour: 230,
+      tooltip: 'Add, or remove allowed type.',
+      helpUrl: 'https://www.youtube.com/watch?v=s2_xaEvcVI0#t=677',
+    });
+  },
+};
+
+export const typeGroupItem = {
+  // Add type.
+  init: function () {
+    this.jsonInit({
+      message0: 'type',
+      previousStatement: null,
+      nextStatement: null,
+      colour: 230,
+      tooltip: 'Add a new allowed type.',
+      helpUrl: 'https://www.youtube.com/watch?v=s2_xaEvcVI0#t=677',
+    });
+  },
+};
+
+const tooltip: Record<string, string> = {
+  null: 'Any type is allowed.',
+  Boolean: 'Booleans (true/false) are allowed.',
+  Number: 'Numbers (int/float) are allowed.',
+  String: 'Strings (text) are allowed.',
+  Array: 'Arrays (lists) are allowed.',
+  CUSTOM: 'Custom type to allow.',
+};
+
+export const type = {
+  init: function (this: Blockly.Block & {customType?: string}) {
+    this.appendDummyInput('TYPE').appendField(
+      new Blockly.FieldDropdown(
+        [
+          ['any', 'null'],
+          ['Boolean', 'Boolean'],
+          ['Number', 'Number'],
+          ['String', 'String'],
+          ['Array', 'Array'],
+          ['other...', 'CUSTOM'],
+        ],
+        (value: string): string => {
+          const customType = 'CUSTOMTYPE';
+          if (value === 'CUSTOM') {
+            if (!this.getField(customType)) {
+              this.getInput('TYPE').appendField(
+                new Blockly.FieldTextInput(this.customType ?? ''),
+                customType,
+              );
+            }
+          } else {
+            this.getInput('TYPE').removeField(customType, true);
+          }
+          return value;
+        },
+      ),
+      'TYPEDROPDOWN',
+    );
+    this.setOutput(true, 'Type');
+    this.setTooltip((): string => {
+      const selectedType = this.getFieldValue('TYPEDROPDOWN');
+      return tooltip[selectedType];
+    });
+    this.setHelpUrl('https://www.youtube.com/watch?v=s2_xaEvcVI0#t=602');
+    this.setStyle('type');
+  },
+  saveExtraState: function () {
+    if (this.getField('CUSTOMTYPE')) {
+      return {customType: this.getFieldValue('CUSTOMTYPE')};
+    }
+  },
+  loadExtraState: function (state: any) {
+    this.customType = state?.customType;
+    if (this.getFieldValue('TYPEDROPDOWN') === 'CUSTOM') {
```

Comment:
Don't rely on checking for the value of the dropdown. Fields are deserialized in an arbitrary order and this may not be correct.